src/System.Net.HttpListener/src/System/Net/HttpListenerRequestUriBuilder.cs

```diff
@@ -0,0 +1,479 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Text;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace System.Net
+{
+    // We don't use the cooked URL because http.sys unescapes all percent-encoded values. However,
+    // we also can't just use the raw Uri, since http.sys supports not only Utf-8, but also ANSI/DBCS and
+    // Unicode code points. System.Uri only supports Utf-8.
+    // The purpose of this class is to convert all ANSI, DBCS, and Unicode code points into percent encoded
+    // Utf-8 characters.
+    internal sealed class HttpListenerRequestUriBuilder
+    {
+        private static readonly Encoding s_utf8Encoding;
+        private static readonly Encoding s_ansiEncoding;
+
+        private readonly string _rawUri;
+        private readonly string _cookedUriScheme;
+        private readonly string _cookedUriHost;
+        private readonly string _cookedUriPath;
+        private readonly string _cookedUriQuery;
+
+        // This field is used to build the final request Uri string from the Uri parts passed to the ctor.
+        private StringBuilder _requestUriString;
+
+        // The raw path is parsed by looping through all characters from left to right. 'rawOctets'
+        // is used to store consecutive percent encoded octets as actual byte values: e.g. for path /pa%C3%84th%2F/
+        // rawOctets will be set to { 0xC3, 0x84 } when we reach character 't' and it will be { 0x2F } when
+        // we reach the final '/'. I.e. after a sequence of percent encoded octets ends, we use rawOctets as 
+        // input to the encoding and percent encode the resulting string into UTF-8 octets.
+        //
+        // When parsing ANSI (Latin 1) encoded path '/pa%C4th/', %C4 will be added to rawOctets and when
+        // we reach 't', the content of rawOctets { 0xC4 } will be fed into the ANSI encoding. The resulting 
+        // string '�' will be percent encoded into UTF-8 octets and appended to requestUriString. The final
+        // path will be '/pa%C3%84th/', where '%C3%84' is the UTF-8 percent encoded character '�'.
+        private List<byte> _rawOctets;
+        private string _rawPath;
+
+        // Holds the final request Uri.
+        private Uri _requestUri;
+
+        static HttpListenerRequestUriBuilder()
+        {
+            s_utf8Encoding = new UTF8Encoding(false, true);
+            s_ansiEncoding = Encoding.GetEncoding(0, new EncoderExceptionFallback(), new DecoderExceptionFallback());
+        }
```

Comment:
It'd be better to initialize these where the fields are declared and remove the explicit cctor.